### ISSUE TITLE
Kotlin based dispatcher file generation support

### ISSUE THREAD

**Issue by hotchemi** [hotchemi/PermissionsDispatcher#320]
Now we have kotlinpoet, it's surely possible to support Kotlin user by generating `***PermissionsDispatcher.kt`! I hope it would solve https://github.com/hotchemi/PermissionsDispatcher/issues/171.

https://github.com/square/kotlinpoet

**Comment by aurae** [comment]
This is going to be pretty involved, but awesome if we can pull it off. There's a lot of things we can leverage with Kotlin as the target of our code generation, so I think it wouldn't necessarily be a direct port of the Java version. For example, we'll probably add the "with check" methods as extension functions to the annotated classes. In the end, we'd ideally have Java and Kotlin targets for the actual file writing, but they share the common logic of the annotation processor that we have in place today.

**Comment by aurae** [comment]
What would be some conveniences that Kotlin can bring to improve the generated code, beyond supporting `internal` in a more appropriate way?

The way I see it, and what I've said before, I'd love to leverage the power of extension functions:

```kotlin
MainActivity.kt:

@RuntimePermissions
class MainActivity : AppCompatActivity() {
    @NeedsPermission(Manifest.permission.CAMERA)
    internal fun accessCamera() {
    }

    fun otherMethod() {
        // Java: "MainActivityPermissionsDispatcher.accessCameraWithCheck(this);"
        this.accessCameraWithCheck()
    }

    override fun onRequestPermissionsResult(rc: Int, permissions: Array<out String>, results: IntArray) {
        // Java: "MainActivityPermissionsDispatcher.onRequestPermissionsResult(this, rc, results);"
        // I'm not satisfied with the method signature here, since it's too similar to the Android one.
        // However, the signature is already pretty long, so I'm open for ideas.
        this.onRequestPermissionsResult(rc, results)
    }
}
```

```kotlin
MainActivityPermissionsDispatcher.kt:

fun MainActivity.accessCamera() {
    // ...
}

fun MainActivity.onRequestPermissionsResult(rc: Int, results: IntArray) {
    // ...
}
```

**Comment by hotchemi** [comment]
Ah cool! I haven't thought about extension functions but it's really neat and make code base intuitive!

**Comment by hotchemi** [comment]
One more thing I just came up with is coroutine. Probably we'll be able to use it for handling async suff!

https://github.com/k-kagurazaka/async-permissions

**Comment by aurae** [comment]
I've seen the "headless-Fragment approach" to handling permissions in another library before, and I think it's quite clever. If we decided to go that route, we wouldn't even need code generation anymore, since any permission request would be handled using a DSL.

Another thing that's holding me back right now is the lack of possibilities to properly assert the generated Kotlin code. Google's compile-testing library doesn't support Kotlin, and because of kapt's stub-based integration with Java annotation processors, it might be quite difficult to ensure the format of generated files (see [here](https://stackoverflow.com/questions/41528471/java-annotation-processor-annotated-kotlin-classes-unit-tests) / [here](https://discuss.kotlinlang.org/t/annotation-processor-unit-testing-inquiry/2204)).

**Comment by agentzheng** [comment]
I can't agree more , I'm going to  to rewrite some of my codes with Kotlin, but the first problem I encounter is I don't know how to  use permission dispatcher in Kotlin. Looking forward to Permission dispatcher for Kotlin.

**Comment by hotchemi** [comment]
@agentzheng in that sense you can use current version with kapt!
This issue is to support more Kotlin-ish way to deal with permissions 😄

**Comment by agentzheng** [comment]
I guess I know what you mean ,but to do so I have to use java while using Kotlin, that is tedious。If we can deal with permission with pure Kotlin ,that would be fantastic.It's a pity that I can do little help due to my poor knowledge. Looking forward  to your next work. By the way ,please forgive my poor English.

**Comment by hotchemi** [comment]
@agentzheng Okay. But I'd say you don't have to write any Java code and Kotlin is 100% interoperate with Java, it means `***PermissionsDispatcher` can be called from .kt file.
This enhancement is kinda experimental and we can't promise accomplishing the work by next week 🦅

**Comment by hotchemi** [comment]
@aurae Thank you for your dedicated research!

> Google's compile-testing library doesn't support Kotlin, 

ic...so do you think it'd be not feasible right?

> "headless-Fragment approach"

I'm curious, which library did you dive into?

**Comment by aurae** [comment]
It would totally be feasible, however it's going to be tough for potential unit tests without an extension to **compile-testing** or **Truth**, which allows us to assert Kotlin code properly.

I was checking out Kagurazaka-san's "async-permissions" library that you shared earlier in the thread. Also I haven't used it so I'm not sure, but I think Dexter does this as well.

**Comment by shiraji** [comment]
We already discussed this type of topic in past but I brought this anyway. I think we need to have some kind of instrument test. No matter what the kind of code generate, the behavior is ok then we are all ok.

**Comment by hotchemi** [comment]
```kotlin
package permissions.dispatcher.samplekotlin

import android.support.v4.app.ActivityCompat
import permissions.dispatcher.PermissionRequest
import permissions.dispatcher.PermissionUtils
import java.lang.ref.WeakReference

private val PERMISSION_SHOWCAMERA = arrayOf("android.permission.CAMERA")

private val REQUEST_SHOWCAMERA = 0

private val REQUEST_SHOWCONTACTS = 1

private val PERMISSION_SHOWCONTACTS = arrayOf("android.permission.READ_CONTACTS", "android.permission.WRITE_CONTACTS")

private class ShowCameraPermissionRequest constructor(target: MainActivity) : PermissionRequest {
    private val weakTarget = WeakReference(target)

    override fun proceed() {
        val target = weakTarget.get() ?: return
        ActivityCompat.requestPermissions(target, PERMISSION_SHOWCAMERA, REQUEST_SHOWCAMERA)
    }

    override fun cancel() {
        val target = weakTarget.get() ?: return
        target.onCameraDenied()
    }
}

fun MainActivity.showCameraWithCheck() {
    if (PermissionUtils.hasSelfPermissions(this, *PERMISSION_SHOWCAMERA)) {
        showCamera()
    } else {
        if (PermissionUtils.shouldShowRequestPermissionRationale(this, *PERMISSION_SHOWCAMERA)) {
            showRationaleForCamera(ShowCameraPermissionRequest(this))
        } else {
            ActivityCompat.requestPermissions(this, PERMISSION_SHOWCAMERA, REQUEST_SHOWCAMERA)
        }
    }
}

fun MainActivity.showContactsWithCheck() {
    if (PermissionUtils.hasSelfPermissions(this, *PERMISSION_SHOWCONTACTS)) {
        showContacts()
    } else {
        if (PermissionUtils.shouldShowRequestPermissionRationale(this, *PERMISSION_SHOWCONTACTS)) {
            showRationaleForContact(ShowContactsPermissionRequest(this))
        } else {
            ActivityCompat.requestPermissions(this, PERMISSION_SHOWCONTACTS, REQUEST_SHOWCONTACTS)
        }
    }
}

private class ShowContactsPermissionRequest constructor(target: MainActivity) : PermissionRequest {
    private val weakTarget = WeakReference(target)

    override fun proceed() {
        val target = weakTarget.get() ?: return
        ActivityCompat.requestPermissions(target, PERMISSION_SHOWCONTACTS, REQUEST_SHOWCONTACTS)
    }

    override fun cancel() {}
}

fun MainActivity.onRequestPermissionsResult(requestCode: Int, grantResults: IntArray) {
    when (requestCode) {
        REQUEST_SHOWCAMERA -> if (PermissionUtils.verifyPermissions(*grantResults)) {
            showCamera()
        } else {
            if (!PermissionUtils.shouldShowRequestPermissionRationale(this, *PERMISSION_SHOWCAMERA)) {
                onCameraNeverAskAgain()
            } else {
                onCameraDenied()
            }
        }
        REQUEST_SHOWCONTACTS -> if (PermissionUtils.verifyPermissions(*grantResults)) {
            showContacts()
        }
        else -> {
        }
    }
}
```

**Comment by hotchemi** [comment]
Umm...I encountered one wired point. Presumably this is a bug of kapt or IntelliJ 🤔 

### prerequisite

- kapt3 supports .kt file generation support.
  - [issue](https://youtrack.jetbrains.com/issue/KT-17883)
  - [commit](https://github.com/jetbrains/kotlin/commit/4a2e40994848883cc7007485dc87d1ab449e45d2)
- then .kt file is supposed to be in `generated/source/kaptKotlin/$sourceSetName`
  - [commit](https://github.com/jetbrains/kotlin/commit/4a2e40994848883cc7007485dc87d1ab449e45d2#diff-5339179fc0c42867e1d6d22697341adcR80)
- and .java file is supposed to be in `generated/source/kapt/$sourceSetName`
  - [commit](https://github.com/jetbrains/kotlin/commit/4a2e40994848883cc7007485dc87d1ab449e45d2#diff-5339179fc0c42867e1d6d22697341adcL75)

---

### problem

Weirdly on IntelliJ or Android Studio, .kt files under the `generated/source/kaptKotlin/$sourceSetName` are not recognized as source file.
But Gradle command doesn't cause any errors.

<img width="1280" alt="screen shot 2017-07-19 at 4 32 45" src="https://user-images.githubusercontent.com/471318/28336311-5fb79422-6c3c-11e7-8188-05b53b2f2579.png">

`generated/source/kapt/$sourceSetName` are recognized as source file correctly.

<img width="1280" alt="screen shot 2017-07-19 at 4 31 11" src="https://user-images.githubusercontent.com/471318/28336388-b5b0880c-6c3c-11e7-8374-1525cfb70e7a.png">

**Comment by aurae** [comment]
That might be an AS-specific issue we should forward to the dev team, so they can make the Android plugin pick up on the `kaptKotlin` sources automatically. I usually go about this with the `idea` plugin & add the source folders to the visibility scope of the IDE myself ([e.g. here - I believe the other property is just called `sourceDirs`](https://github.com/aurae/android-junit5/blob/master/android-junit5/build.gradle#L47)), but that would be an annoying workaround for our users in the long run.

Thank you for the amazing work on this so far!

**Comment by hotchemi** [comment]
opened a issue https://youtrack.jetbrains.com/issue/KT-19097!

**Comment by dimsuz** [comment]
FYI: Not recognizing `kaptKotlin` source set was was partly due to a bug in AS-3.2.0/3.3.0. I reported it and it got addressed in 3.3.1.

Writing because I noticed you are still using the [workaround](https://github.com/permissions-dispatcher/PermissionsDispatcher/blob/master/processor/src/main/kotlin/permissions/dispatcher/processor/PermissionsProcessor.kt#L75-L78) in the code.

Maybe it can be removed now.

There's still an issue with this even in 3.3.1, but it occurs only in presence of a new compiler flag, I also [filed it](https://issuetracker.google.com/issues/124024529) and it is said to be fixed in upcoming release.
